CesiumIonClient/include/CesiumIonClient/Geocoder.h

```diff
@@ -0,0 +1,131 @@
+#pragma once
+
+#include "CesiumGeospatial/Cartographic.h"
+#include "CesiumGeospatial/GlobeRectangle.h"
+
+#include <CesiumUtility/Math.h>
+
+#include <string>
+#include <variant>
+#include <vector>
+
+namespace CesiumIonClient {
+
+/**
+ * @brief The supported types of requests to geocoding API.
+ */
+enum GeocoderRequestType {
+  /**
+   * @brief Perform a full search from a complete query.
+   */
+  Search,
+
+  /**
+   * @brief Perform a quick search based on partial input, such as while a user
+   * is typing.
+   * The search results may be less accurate or exhaustive than using {@link GeocoderRequestType::Search}.
+   */
+  Autocomplete
+};
+
+/**
+ * @brief The supported providers that can be accessed through ion's geocoder
+ * API.
+ */
+enum GeocoderProviderType {
+  /**
+   * @brief Google geocoder, for use with Google data.
+   */
+  Google,
+
+  /**
+   * @brief Bing geocoder, for use with Bing data.
+   */
+  Bing,
+
+  /**
+   * @brief Use the default geocoder as set on the server. Used when neither
+   * Bing or Google data is used.
+   */
+  Default
+};
+
+/**
+ * @brief A single feature (a location or region) obtained from a geocoder
+ * service.
+ */
+struct GeocoderFeature {
+  /**
+   * @brief The user-friendly display name of this feature.
+   */
+  std::string displayName;
+
+  /**
+   * @brief The region on the globe for this feature.
+   */
+  std::variant<CesiumGeospatial::GlobeRectangle, CesiumGeospatial::Cartographic>
+      destination;
+
+  GeocoderFeature(
+      std::string& displayName_,
+      CesiumGeospatial::GlobeRectangle& destination_)
+      : displayName(displayName_), destination(destination_) {}
+  GeocoderFeature(
+      std::string& displayName_,
+      CesiumGeospatial::Cartographic& destination_)
```

Comment:
```suggestion
      const std::string& displayName_,
      const CesiumGeospatial::Cartographic& destination_)
```

Reply:
Alternatively, you could do this:

```cpp
GeocoderFeature(
      std::string displayName_,
      const CesiumGeospatial::Cartographic& destination_)
      : displayName(std::move(displayName_)), destination(destination_) {}
```

Which would be slightly more efficient because you can optionally move the string into the constructor (moving a GlobeRectangle isn't any different from copying it).

Slightly more efficient still would be two constructors, but that's almost certainly not worth the noise:

```cpp
GeocoderFeature(
      std::string&& displayName_,
      const CesiumGeospatial::Cartographic& destination_)
      : displayName(std::move(displayName_)), destination(destination_) {}

GeocoderFeature(
      const std::string& displayName_,
      const CesiumGeospatial::Cartographic& destination_)
      : displayName(displayName_), destination(destination_) {}
```

Reply:
Another possibility here is to just get rid of the constructors entirely for these simple value-holding types. Users could still create instances with almost the exact same syntax, except they'd have to use curly braces.